arch/arm/core/thread.c

```diff
@@ -298,3 +298,23 @@ u32_t z_check_thread_stack_fail(const u32_t fault_addr, const u32_t psp)
 	return 0;
 }
 #endif /* CONFIG_MPU_STACK_GUARD || CONFIG_USERSPACE */
+
+#if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)
+int z_arch_float_disable(struct k_thread *thread)
+{
+	if (thread != _current) {
+		return -EINVAL;
+	}
+
+	if (z_is_in_isr()) {
+		return -EINVAL;
+	}
+
+	/* Disable all floating point capabilities for the thread */
+
+	thread->base.user_options &= ~K_FP_REGS;
+
+	__set_CONTROL(__get_CONTROL() & (~CONTROL_FPCA_Msk));
```

Comment:
Do those two lines need locking?  Naively looking at them I'd say those two lines need to happen atomically.

Reply:
This is a really interesting question, @andyross :)
Few comments
- `z_arch_float_disable` cannot be called in an ISR for the ARM architecture (this is documented and an ASSERT has also been added to verify this in run-time).

- ISRs cannot use the FP context - this is also part of Zephyr spec. So ISRs won't play with the CONTROL register.

- An IRQ handler is supposed to restore the CONTROL register to its initial state, before a preempting interrupt. 

- a thread context-switch -out and -in is _also_ supposed to restore the CONTROL register to its initial state, before swap-out. 

Therefore, I believe that we are safe here, even if the thread is swapped-out and -in right during the read-modify-write operation on the special-purpose CONTROL register.

Any different opinions, Andy?

Reply:
As long as you looked at the problem and decided it's safe, then I'm fine with that.

The situation I was thinking of was the OS taking an interrupt between those two lines, and then making assumptions about the saved stack frame based on the flag, but the actual state being different because the hardware state didn't match.   For example interrupt exit for a "FPU" task might be different from a "non-FPU" task.  But the interrupt was actually taken on a process that was still using the FPU.

Reply:
Oh, I forgot to reply on this, too, @andyross , sorry :) 

In brief, `K_FP_REGS` is **_currently_** not used in the ARM stack. It was added, here, in order to align the implementation with the x86 k_float_disable() function. So there is no need for interrupt locking here. But, I say, "currently", because we are planning to use this option-flag in when doing the work for #16360.

I am anticipating that even after #16360, there will be no need for locking, here, either. The reason is that the option-flag K_FP_REGS will only be used to decide the width of the MPU stack guard. So, the worst case scenario in the situation you described is:
1. a k_float_disable() clears K_FP_REGS
2. an interrupt causing a SWAP occurs right between the two lines
3. before the thread is swapped-in again, it, now, has a shorter MPU stack guard (as non-FP capable)

However, as the _first_ thing that the thread does is to modify the CONTROL register (without using any stack memory), this is safe.

But I anyways added a comment in #16360, to revisit the need for locking here. Thanks!